Implemented Commander variant of Adventure Mode

jgroehl opened the pull request:
There was an open project on the Wiki for this - here is an implementation for it.
I tested it through some play and it works and at least for me is a lot of fun!
Through a short discussion with @kazyfreak [here](https://github.com/jgroehl/forge/issues/1), I reverted any enforced restrictions I wanted to put in place initially. 
At the end of the day, the player can now have a commander, is supplied with a starting commander + deck, and has full deckbuilding flexibility.

Please let me know if you think and if this could be included into `main`!

Summary of code changes:
- New Adventure Scene: 
  - Added "Commander" option.
  - Supplies player with 15 (random) startup commanders.
  - Communicates the chosen commander to the world creation.
  - Added a small card preview (hijacking the RewardActor) so the player can see the chosen card.
- World Creation
  - logical flow fork if a commander has been chosen
- Adventure Player
  - Now stores a flag whether Adventure is played in commander mode.
- Config (most of the added lines of code)
  - added a function to heuristically add a somewhat functional random deck pile for the chosen starting commander.
- Duel Scene
  - Started in commander mode if adventure is in the commander variant
- Reward Screen
  - Added a convenience function where cards that are already possessed are flagged for autosell automatically.

Jetz72 commented on forge-gui-mobile/src/forge/deck/FDeckEditor.java:
```diff
@@ -1084,7 +1085,7 @@ public boolean isLimitedEditor() {
     }
     public boolean isCommanderEditor() {
-        return editorConfig.hasCommander();
+        return AdventurePlayer.current().isCommanderMode() || editorConfig.hasCommander();
```
I think this will crash the deck editor outside Adventure mode. Either way, Adventure specific stuff should be in `AdventureDeckEditor`.

Jetz72 commented on forge-gui-mobile/src/forge/adventure/player/AdventurePlayer.java:
```diff
@@ -334,6 +348,14 @@ public void setColorIdentity(ColorSet set) {
+    private Deck newDeck(String deckname){
+        if (commanderMode) {
+            Deck n_deck = new Deck(deckname);
+            n_deck.putSection(DeckSection.Commander, new CardPool());
```
Was something breaking if you didn't manually initialize this? Most places use `Deck::getOrCreate(DeckSection)` and should be able to handle its absence.

Jetz72 commented on forge-gui-mobile/src/forge/adventure/scene/DuelScene.java:
```diff
@@ -197,6 +197,8 @@ public void enter() {
         if (eventData != null && eventData.eventRules != null) {
             mainGameType = eventData.eventRules.gameType;
+        } else if (AdventurePlayer.current().isCommanderMode()){
+            mainGameType = GameType.Commander;
```
You'll also want to update `DuelScene::applyAdventureDeckRules` to handle missing commanders, illegal commanders, mismatched color identity, and any other commander deck rules.

jgroehl replied:
That all works through line 211 here, since the DeckFormat is adjusted properly in the Enum:

        applyAdventureDeckRules(mainGameType.getDeckFormat());

jgroehl replied:
Regarding the balancing: From the playtesting I have done so far, I find the experience much more rewarding, actually! 100 singletons gives you a lot of variance - it does not become a given that you can defeat any boss, just because you assemble the first cards for an infinite loop etc.

It is a much more massive undertaking assembling a good multi-colour commander deck, since you need to travel around a lot chasing for cards that work!

I have also not found it to be problematic that the opponents do not use a commander.

Jetz72 replied:
Upon review I think that method is meant to be gated behind a check for `FModel.getPreferences().getPrefBoolean(FPref.ENFORCE_DECK_LEGALITY)`. Anyone who wants to play with freeform decks should be able to just disable that and then do so.

But for everyone else, yes I think it'd be sensible to make sure their deck meets the format criteria. We could maybe ignore the case where they have no commander since that's just a strictly worse deck 99.999% of the time. But as for the color identity rule and illegal commanders, I'd rather not leave it to the honor system to enforce it.

Jetz72 reviewed (commented):
The biggest hurdle here is that Adventure mode isn't really balanced for Commander. Items and encounters and bosses are designed with 40-60 card constructed decks with limited inventory in mind. How thoroughly have you tested it with balance in mind?

Jetz72 commented on forge-gui-mobile/src/forge/adventure/scene/AdventureDeckEditor.java:
```diff
@@ -896,12 +916,20 @@ public void onClose(final Consumer<Boolean> canCloseCallback) {
+            if (AdventurePlayer.current().getAdventureMode() == AdventureModes.Commander)
+                deckError = GameType.Commander.getDeckFormat().getDeckConformanceProblem(getDeck());
+            else
+                deckError = GameType.Adventure.getDeckFormat().getDeckConformanceProblem(getDeck());
```
I think what you'd want to do is acquire `DeckFormat.Commander` via `AdventureEditorConfig.getDeckFormat`, then just simplify this to `getEditorConfig().getDeckFormat().getDeckConformanceProblem(...`. The editor config's deck format is used for various other bits of editor logic.

jgroehl commented:
@Jetz72 @tool4ever Thanks for your feedback again 🙏.
I've implemented your suggestions, making the code edits quite minimal. The only larger chunk of code is the convenience function to automatically toggle cards for autosale on the reward screen that the player already had in their collection.
I've also cleaned up unused imports and accidentally introduced line breaks, etc. 
Would really appreciate another review!
Cheers 😄

tool4ever commented:
`mvn test` for the full suite
single ones depending on how your IDE works...

tool4ever commented on forge-gui-mobile/src/forge/adventure/scene/NewGameScene.java:
```diff
@@ -446,10 +533,31 @@ private void showModeHelp() {
-                summaryText.append("Mode: Chaos\n\nYou (and all enemies) will receive a random preconstructed deck.\n\nWarning: This will make encounter difficulty vary wildly from the developers' intent");
+                summaryText.append("""
+                        Mode: Chaos
+                        
```
does hard converting these linebreaks work for all platforms?

jgroehl replied:
AFAIK, they are part of the Java language specification since Java 15 and thus not platform-dependent.
But this is an "unnecessary" alteration regarding the objective of this issue, so I am happy to revert these and align the added text.

jgroehl replied:
I've reverted to the way it was before, just in case.

Jetz72 commented on forge-gui-mobile/src/forge/adventure/scene/NewGameScene.java:
```diff
@@ -233,6 +288,31 @@ public void act(float delta) {
+            return isLegendaryCreature && sharesColorInIdentity;
+
+        };
+        List<PaperCard> commanderCandidates = new ArrayList<>(cardDb.streamUniqueCardsNoAlt().filter(commander_filter).toList());
```
I worry that offering random commander candidates before the adventure starts just encourages rerolling by backing out and starting over. Perhaps a fixed list of like 2 or 3 starter commanders with simple themes for each mono-color?

kazyfreak replied:
How I rolled the version I'm testing (since its got a slightly different scope) is that I hard coded the 5 single color, 10 duels and triomes into the color picker (adding in symbols etc). Then linked those to a precon. They are staples, easy to adjust offers a few alternate commanders to build on, and as balanced as your going to get before players start customizing. That gives them something to start with that is viable, but tons of upgrades and they can pick colors or themes they are interested in if they want to go find a different commander elsewhere.

I also plan on coding enemies to have a % chance to drop what commander they are using. That way players can farm them if they want.

Again different game but I thought I'd offer my two cents.

I did have a slight issues I had to rebuild since the commander match wasn't an adventure mode. Some logic changes I did to my own forked version (Just things to keep an eye on)

Added Commander deck editor support with validation | AdventureDeckEditor.java (lines 67-91, 916-923) |
Added Commander detection to adventure battles | DuelScene.java (lines 201-204) |
Simplified detection logic + added debug output (deck editor not showing tab) | AdventureDeckEditor.java (lines 67-94) |
Fixed validation persistence + added commander menu items | AdventureDeckEditor.java (lines 614-620, 919-925) |
Fixed GameRules to use detected mainGameType for match logic | DuelScene.java (line 370)
Fixed match exit screen to use AdventureWinLose for all GameTypes in adventure mode | ViewWinLose.java (lines 89-92) |Fixed IndexOutOfBoundsException when loading maps with missing enemies (biome bounds check) | MapStage.java (lines 490-496) | *This is because we're not using all the common enemies bc they aren't commander*

jgroehl replied:
Great, thanks for the quick feedback!
Based on your suggestions, I would tend to agree with @kazyfreak's opinion.
To start off, let's offer a single commander precon for each monocolor, but implemented in a way that it is as simple as adding more precons to eventually enable more choices in the future.

-> This way, we enable a balanced start to the game, and encourage the player to build a collection during the playthrough for building stronger decks!
-> game design-wise, I'd want to go for weaker decks that rely on playing on-curve creatures to win the game, and restrict the 99 to commons/uncommons. -> from the beginning, upgrades will feel meaningful

@kazyfreak thanks for sharing the changes you made - that is mostly in line with the changes I made in the PR! In the proposed version, enemies remain untouched - they play their usual decks, and it is only the player who has access to the command zone.

jgroehl replied:
@kazyfreak do you mind sharing the deck lists for the monocolor starter decks you built?

jgroehl commented:
I believe this is now good to go. I've playtested a bit more, and it all seems to make sense.
Thanks everyone for the feedback so far - happy to make more changes if you find more things out of place!

The changes are now quite minimal to enable the adventure player to run the playthrough with a commander deck.
The only really optional change is the autosell toggle for duplicates.
The starter decks can all win games against the early enemies, but have no chance against the bosses -> I think this is a good balance to get this game mode started!

In the future, it would be great to add specific features to this game mode, i.e. difficulty-specific starter decks, special enemies that also have commanders, dedicated shops for commander-related cards, 4-player EDH tournaments in the capital, EDH-specific gear rewards, changes to rewards in EDH settings (no card drops more than once), etc, etc :) Happy to support these developments if you think these would make sense!

jgroehl commented:
Since I have changed the data save by saving and loading an additional flag, it would be fantastic if someone else apart from me could test that the changes do not break the save file when loading. They did not for me for the save file I created before the changes, but it would be great to double-check!!

Zara7hus7ra commented:
sol ring in starter deck seems a little bit broken?
